cmd/nri-kubernetes/main.go

```diff
@@ -0,0 +1,199 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/newrelic/infra-integrations-sdk/integration"
+	"github.com/newrelic/infra-integrations-sdk/log"
+	"github.com/sethgrid/pester"
+	"k8s.io/apimachinery/pkg/version"
+
+	"github.com/newrelic/nri-kubernetes/v2/internal/config"
+	"github.com/newrelic/nri-kubernetes/v2/internal/discovery"
+	"github.com/newrelic/nri-kubernetes/v2/src/apiserver"
+	"github.com/newrelic/nri-kubernetes/v2/src/client"
+	"github.com/newrelic/nri-kubernetes/v2/src/ksm"
+	"github.com/newrelic/nri-kubernetes/v2/src/metric"
+	"github.com/newrelic/nri-kubernetes/v2/src/scrape"
+	"github.com/newrelic/nri-kubernetes/v2/src/sink"
+	"github.com/newrelic/nri-kubernetes/v2/src/storage"
+)
+
+const defaultLabelSelector = "app.kubernetes.io/name=kube-state-metrics"
+
+type clientsCluster struct {
+	k8s                 client.Kubernetes
+	ksm                 ksm.Client
+	api                 apiserver.Client
+	endpointsDiscoverer discovery.EndpointsDiscoverer
+	servicesDiscoverer  discovery.ServiceDiscoverer
+}
+
+func main() {
+	c := config.LoadConfig()
+	logger := log.NewStdErr(c.Verbose)
+
+	k8s, err := client.NewKubernetes(true)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	apiClient := apiserver.NewFileCacheClientWrapper(apiserver.NewClient(k8s), client.DiscoveryCacherConfig{
+		TTL:       3 * time.Hour,
+		TTLJitter: 50,
+		Storage:   storage.NewJSONDiskStorage("/var/cache/nr-kubernetes/apiserverK8SVersion"),
+	})
+
+	ksmDiscoverer, err := getDiscoverer(c, k8s, logger)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	servicesDiscoverer := discovery.NewServicesDiscoverer(k8s.GetClient())
+
+	ksmClient, err := ksm.NewKSMClient(logger)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	clients := clientsCluster{
+		k8s:                 k8s,
+		ksm:                 ksmClient,
+		api:                 apiClient,
+		endpointsDiscoverer: ksmDiscoverer,
+		servicesDiscoverer:  servicesDiscoverer,
+	}
+
+	if err := run(c, logger, clients); err != nil {
+		log.Fatal(err)
```

Comment:
For the future `main`, it might be good to have individual objects with their own run functions.